n3fit/src/n3fit/backends/keras_backend/operations.py

```diff
@@ -134,6 +141,8 @@ def numpy_to_input(numpy_array, no_reshape=False, name=None):
         batched_array = np.expand_dims(numpy_array, 0)
         batch_size = 1
         shape = numpy_array.shape
+    if custom_shape is not None:
+        shape = custom_shape
```

Comment:
The reason was that without this, several layers get passed multiple shapes, e.g. (1, None, 1) for the grid, (1, 2000, 1) for the integration grid and (1, 1, 1) for the x=1 input. Because of this, in the model.summary the shapes are not displayed, but rather something like [?]. I apply it to the latter two inputs to make them consistent. It was explained at one of the two points it was used, but not very clearly, so I'll add it to the docstring.

On how to clean up the `numpy_to_input` function, I see no harm in doing this always, replacing whatever the number of gridpoints is with None. 
Also, the other boolean argument `no_reshape` is not used anywhere as far as I can tell (e.g. barring any reportengine magic). This means that everything passed to it is without batch dimension, so we can get rid of both the `custom_shape` and the `no_reshape` arguments and the if statement entirely, and set the shape to that of the input with the 0 index replaced by None. Do you agree? (I'll have to test ofcourse, still/again don't have a working local installation so it's cumbersome..)

Reply:
This is implemented in [this commit](https://github.com/NNPDF/nnpdf/pull/1734/commits/80ef5d72a5bf1c2476ba1f0bd1374b2d7239c700)

Reply:
Ah, I hadn't realised it affects the model.summary, but I agree with everything you say here. There is no reportengine magic happening in this part of the code since reportengine isn't even aware of the existence of this function. I do understand that what reportengine does may seem a bit like black magic but it works a lot like [pytest fixtures ](https://docs.pytest.org/en/6.2.x/fixture.html), if you're familiar with that.